### ISSUE TITLE
more example code please... 

### ISSUE THREAD

**Issue by peterthorpe5** [BojarLab/glycowork#20]
Dear Glycowork,

Thanks you for your software. I have been trying to work through this so I can predict the binding capacity of some lectins from their sequence using lectin_oracle_flex, whoever the example code ins https://bojarlab.github.io/glycowork/examples.html#example2 is not enough for me to work with in order to do it myself. 

Would you kindly please expand your example for the lecting section?

kind regards,

Peter Thorpe

**Comment by pvalle6** [comment]
Hey Peter,

If you would like to use lectin_oracle_flex, you should import all of the dependencies of glycowork and the relevant glycowork files. 
In addition, you should create a library of glycans or use the provided glycowork glycan set to predict your lectin against and import a prepared model or prepare your own model. 

```
model =  prep_model(model_type, num_classes, libr = None,
               trained = False):
    """wrapper to instantiate model, initialize it, and put it on the GPU\n
    | Arguments:
    | :-
    | model_type (string): string indicating the type of model
    | num_classes (int): number of unique classes for classification
    | libr (list): sorted list of unique glycoletters observed in the glycans of our dataset\n
    | Returns:
    | :-
    | Returns PyTorch model object
    """
```

The prediction can be returned using the function 
```
get_lectin_preds(prot, glycans, model, prot_dic = {}, background_correction = False,
                     correction_df = None, batch_size = 128, libr = None, sort = True,
                     flex = False):

  |prot (string): protein amino acid sequence
  | glycans (list): list of glycans in IUPACcondensed
  | model (PyTorch object): trained LectinOracle-type model
  | prot_dic (dictionary): dictionary of type protein sequence:ESM1b representation
  | background_correction (bool): whether to correct predictions for background; default:False
  | correction_df (dataframe): background prediction for (ideally) all provided glycans; default:V4 correction file
  | batch_size (int): change to batch_size used during training; default:128
  | libr (list): sorted list of unique glycoletters observed in the glycans of our dataset
  | sort (bool): whether to sort prediction results descendingly; default:True
  | flex (bool): depends on whether you use LectinOracle (False) or LectinOracle_flex (True); default:False\n`

```
`model = prep_model(LectinOracle_flex,1, trained = True)
`
`prediction = get_lectin_preds(prot,glycans,model, dictionary, flex = True) #dictionary may not be necessary for flex`

This should get you started.

**Comment by peterthorpe5** [comment]
thank you for your reply. I will look into this in detail and let you know how I get on. Thank you again!!

**Comment by Bribak** [comment]
Thanks all for asking & answering!:-)

Currently traveling; thus the delay on my part.
Good suggestion; Long-term, we'll add another example to the documentation in v0.6 (~October or November). Short-term, let me know if you need anything more / more specific until then and I'll try to help you here.

**Comment by peterthorpe5** [comment]
Dear Glycowork

Could you please give me an example of what to put in to the  "prot_dic (dictionary)" when calling the "get_lectin_preds" function?

    prot_dic (dictionary): dictionary of type protein sequence:ESM1b representation

    get_lectin_preds(prot, glycans, model, **prot_dic = {}**, background_correction = False,
                     correction_df = None, batch_size = 128, libr = None, sort = True,
                     flex = False):

could it be as simple as protein_dicr = {"QWERTFVCF": "Lectin_name"}

regards,

Pete

**Comment by Bribak** [comment]
Dear Pete,

sure!
The prot_dic has protein sequences as keys and their ESM-1b representation as values. You can get the ESM-1b representation of a sequence outside of glycowork by running:

```
!pip install fair-esm
import esm
model, alphabet = esm.pretrained.esm1b_t33_650M_UR50S()
```

and then use our function `glycowork.ml.inference.get_esm1b_representations`, which converts a list of protein sequences to the prot_dic that you can use as an input for get_lectin_preds.

Note that the above might take a second, as it has to download the model weights of the, fairly huge, ESM-1b model.
An alternative could be to use LectinOracle_flex, where you can directly use protein sequences without having a prot_dic (though with slightly worse performance).

Let me know if anything seems unclear to you!
Thanks & Best,
Daniel

**Comment by peterthorpe5** [comment]
Thank you very  much for your help and quick reply!

"..which converts a list of protein sequences to the prot_dic that you can use as an input for get_lectin_preds"

If I was only trying to analyse one sequence can this work? - As it expects a list of sequences ... what would you use as the list?

**Comment by Bribak** [comment]
That should be no problem, it's just that the datatype list is expected but you can easily make a list of size 1; i.e., ["QWERTFVCF"] in your example

**Comment by Bribak** [comment]
Hi Rhys,

happy to help!
This line:
`glycans = [glycowork.glycan_data.loader.df_glycan]`
makes a list of a dataframe (namely df_glycan). What you want/need to do is to have a list of glycan sequences instead.
So the minimal change would be 
`glycans = df_glycan.glycan.values.tolist()`
if you import df_glycan

I would however rather recommend importing df_species instead (it's also in glycowork.glycan_data.loader) and do
`glycans = list(set(df_species.target.values.tolist()))`
as df_glycan contains many unphysiologial glycan entries that are imported from GlyTouCan and that will need to be manually curated at some point. df_species is already curated but contains duplicates (same glycan in different species), thus making it a set will make it more efficient.

Hope that helps, let me know if there are other questions!
Daniel

**Comment by Bribak** [comment]
Also (just spotted this):
`model = fp+'LectinOracle_565.pt'`
You're loading in a trained LectinOracle version that is not the flex version (so even after fixing the above you'll get an error). In general, using a filepath as an input will not work with the functions. You'll need to use prep_model to initialize your chosen model.
In your case that will be something along the lines of:
`model = prep_model(model_type = "LectinOracle_flex", num_classes = 1, trained = True)`
Then, you can pass that model object to the prediction function

**Comment by rhysDevans97** [comment]
Hi Daniel,
That has been super helpful; looks like that is working, so many thanks!

**Comment by rhysDevans97** [comment]
Hi again,
I have tried using the ESM1b model instead but get a RunTime error from the esm script, though I do not get this issue with the flex version. I am using PyTorch 1.8.1 if that is relevant to Tensor. What could be causing this issue?
```
my_seq = "QWERTFVCF"
esm1b = True
if esm1b:
  !pip install fair-esm
  import esm
  model, alphabet = esm.pretrained.esm1b_t33_650M_UR50S()
  my_seq_esm = get_esm1b_representations([my_seq], model = model, alphabet = alphabet)
  prediction = get_lectin_preds(prot = my_seq, glycans = glycans, model = model, prot_dic = my_seq_esm, background_correction = False, correction_df = None, batch_size = 128, libr = None, sort = True, flex = False)

---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
[<ipython-input-4-c13b6d0249f2>](https://localhost:8080/#) in <module>
     17   my_seq_esm = get_esm1b_representations([my_seq], model = model, alphabet = alphabet) # Can be used by LectinOracle
     18   prediction = get_lectin_preds(prot = my_seq, glycans = glycans, model = model, prot_dic = my_seq_esm, background_correction = False,
---> 19                      correction_df = None, batch_size = 128, libr = None, sort = True)
     20 else:
     21   model = prep_model(model_type = "LectinOracle_flex", num_classes = 1, trained=True)

3 frames
[/usr/local/lib/python3.7/dist-packages/esm/model/esm1.py](https://localhost:8080/#) in forward(self, tokens, repr_layers, need_head_weights, return_contacts)
    115 
    116     def forward(self, tokens, repr_layers=[], need_head_weights=False, return_contacts=False):
--> 117         if return_contacts:
    118             need_head_weights = True
    119 

RuntimeError: Boolean value of Tensor with more than one value is ambiguous
```

**Comment by Bribak** [comment]
Hmm, I suspect this might indeed by a PyTorch version issue (could be that the stored esm1b version is tied to a higher torch version). If I run the snippet below (PyTorch 1.12) it runs fine, though I haven't seen at which torch version it would start to break (LectinOracle was only trained on torch >= 1.9).
Probably no easy way out of upgrading your torch version, if the incompatibility is on the esm side. Hope that helps a bit!

!pip install fair-esm
import esm
!pip install glycowork
import glycowork
from glycowork.ml.inference import get_esm1b_representations, get_lectin_preds
from glycowork.ml.models import prep_model
model, alphabet = esm.pretrained.esm1b_t33_650M_UR50S()
rep = get_esm1b_representations(["QWERTFVCF"], model, alphabet)
leor = prep_model("LectinOracle", 1, trained=True)
get_lectin_preds("QWERTFVCF", ['Neu5Ac(a2-6)GalNAc', 'Gal(b1-4)Glc'], leor, rep)

**Comment by rhysDevans97** [comment]
Ah yes I was assigning the wrong variable for the model in get_lectin_preds. Thanks for providing the snippet

**Comment by peterthorpe5** [comment]
Dear Glycowork,  

Thank you for all your help so far! I am now onto the next step where I would like to make a heatmap of my results and get the get_pvals_motifs:

    make_heatmap(glycan_binding, mode = 'motif', feature_set = ['exhaustive'], datatype = 'response', yticklabels = 1,
            xticklabels = False)
    results = get_pvals_motifs(glycan_binding, feature_set = ['exhaustive'], multiple_samples = True)


I have results as such from the:
        get_lectin_preds(prot, glycans, model, prot_dic = prot_dic, background_correction = False,
                     correction_df = None, batch_size = 128, libr = None, sort = True,
                     flex = True)

	motif	pred
2	GalNAc(a1-4)GlcNAcA(a1-4)[GlcN(b1-7)]Kdo(a2-5)...	-0.380128
0	Man(a1-3)[Man(a1-6)][Xyl(b1-2)]Man(b1-4)GlcNAc...	-0.001911
3	Man(a1-3)[Man(a1-6)][Xyl(b1-2)]Man(b1-4)GlcNAc...	0.010994
1	Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-3)Man(a1-6)...	0.043963

What I don't understand is how to get the "glycan_binding" to input into the get_pvals_motifs and heatmap functions. Looking at your test scripts:

    test_df = pd.DataFrame({'glycan':glycans, 'binding':label})
    print(get_pvals_motifs(test_df, 'glycan', 'binding'))

How do I/ what do I use for label, and what is the most sensible to assign what ever I need to label?

kind regards,

Peter Thorpe

**Comment by Bribak** [comment]
Sure, happy to help!

`get_pvals_motifs` basically needs two columns: (i) containing your glycan sequences & (ii) containing your binding or binding predictions. How you name them is up to you, just make sure you provide the column names as glycan_col_name (i) and label_col_name (ii) in the function call. In your case that would be "motif" (i) and "pred" (ii)
For (z-score transformed) binding data, I would just go with the default threshold set in the function. But for binding predictions, I would probably start with a thresh = 0, as the distribution of binding prediction values is different.
The output will then be a test of which motifs are statistically enriched in glycans with predicted binding > 0 

`make_heatmap` is slightly less flexible and expects that you have multiple samples that you want to compare/cluster on the heatmap. In this case, the input would be a Sample X Glycan matrix, with the binding / binding predictions as values. Once you have that, you should be able to run the code like for instance shown in the example of the documentation.
In theory you could still do this with one sample (by having a single row, with the glycans as columns) but that would likely not be very informative.

Hope that helps!

**Comment by peterthorpe5** [comment]
Thank you very much for your response, I will try this now. Thanks again!!!

**Comment by peterthorpe5** [comment]
Im sorry to ask this, but would you be kind enough to please give me some guidance on getting it into a dataframe? I have tried:

    test_df = pd.DataFrame({'glycan':data.iloc[:,0], 'binding':data.iloc[:,1]})

then:
    results = get_pvals_motifs(test_df, feature_set = ['exhaustive'], multiple_samples = True)

which returns loads of errors:

 DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  out_matrix[label_col_name] = temp
... and TypeError: '>' not supported between instances of 'str' and 'float'

regards,

Pete

**Comment by Bribak** [comment]
It seems that you already have the data in a dataframe; you could just that as input as well.
But I think the main issue is a column name mismatch. If you look at the function definition of `get_pvals_motifs`, you'll see that it has default values for column names
`glycan_col_name = 'glycan', label_col_name = 'target'`
So you will need to either have the same column names in your dataframe or provide different column names when using the function, such as 
`results = get_pvals_motifs(test_df, label_col_name = 'binding', feature_set = ['exhaustive'], multiple_samples = True)`

Also, note that you're currently using `multiple_samples = True` but the test_df you construct only has one sample. If that is your final dataframe, probably easier to leave it at False instead.

Finally, the "highly fragmented" thing is "just" a warning for efficiency reasons; it's already fixed on the dev branch to not occur anymore, in preparation for the version update.
And hopefully the type error will be fixed by the above (fingers-crossed). If the error still persists, probably easiest to show me how test_df looks like when printed; just to make sure there's not format mismatch

Best,
Daniel

**Comment by peterthorpe5** [comment]
Thank you again for your help. Is it ok for me to share an ipyn notebook with you, as I am still having issues with this?

regards, Pete

**Comment by Bribak** [comment]
Sure, go ahead!

**Comment by peterthorpe5** [comment]
[glycowork_gpu.zip](https://github.com/BojarLab/glycowork/files/10012315/glycowork_gpu.zip)

I had to zip it for github to allow it .. please find it attached. Thank you.

**Comment by Bribak** [comment]
Thanks! Before I run it etc; could you tell me what the issue is? I only see warnings in the output - not errors, so that's not an issue per se. Did you try printing results to see whether it looks like you'd expect?

**Comment by peterthorpe5** [comment]
:) my mistake. Yes, ok so it did run. The results were, all corr_pval = 1:
                  motif  pval  corr_pval
0                   Fuc   1.0        1.0
1                GalNAc   1.0        1.0
2                  GlcN   1.0        1.0
3                GlcNAc   1.0        1.0
4               GlcNAcA   1.0        1.0
5                GlcOPN   1.0        1.0
6                   Kdo   1.0        1.0
7                   Man   1.0        1.0
8                   Xyl   1.0        1.0
9                  a1-2   1.0        1.0
10                 a1-3   1.0        1.0
11                 a1-4   1.0        1.0
12                 a1-6   1.0        1.0
13                 a2-4   1.0        1.0
14                 a2-5   1.0        1.0
15                 a2-6   1.0        1.0
16                 b1-2   1.0        1.0
17                 b1-4   1.0        1.0
18                 b1-6   1.0        1.0
19                 b1-7   1.0        1.0
20                 bond   1.0        1.0
21      Fuc(a1-3)GlcNAc   1.0        1.0
22  GalNAc(a1-4)GlcNAcA   1.0        1.0
23        GlcN(b1-7)Kdo   1.0        1.0
24   GlcNAc(b1-4)GlcNAc   1.0        1.0
25   GlcNAc(bond)GlcNAc   1.0        1.0
26     GlcNAcA(a1-4)Kdo   1.0        1.0
27   GlcOPN(b1-6)GlcOPN   1.0        1.0
28         Kdo(a2-4)Kdo   1.0        1.0
29         Kdo(a2-5)Kdo   1.0        1.0
30      Kdo(a2-6)GlcOPN   1.0        1.0
31         Man(a1-2)Man   1.0        1.0
32         Man(a1-3)Man   1.0        1.0
33         Man(a1-6)Man   1.0        1.0
34      Man(b1-4)GlcNAc   1.0        1.0
35         Xyl(b1-2)Man   1.0        1.0

**Comment by Bribak** [comment]
Okay, good! Yes, those values are expected if you didn't change the thresh value in get_pvals_motifs (it's default is 1.645, so if no prediction is over that value, then nothing can be significant of course;-) )

**Comment by peterthorpe5** [comment]
Thank you again for all your help, I have adjusted the thresh as you suggested! Much appreciated!!!

thank you, Pete

**Comment by Bribak** [comment]
Fixed with #22
